### DISCUSSION TITLE
Not able to load viewer. Error Unsupported access token format

### DISCUSSION THREAD

**Discussion by mattxerini** [iTwin/itwinjs-core#5352] [Q&A] [answered]
Good morning, 

I keep seeing this error when I try to launch the viewer within my app.
The token should be correct as it was used previously, although we were using a very old version 2.x 

<img width="605" alt="image" src="https://user-images.githubusercontent.com/123177586/230049457-89f699ae-efd0-45e4-9eb4-3409633a6454.png">

Could somebody help with this?

**Comment by pmconne** [comment]
> we were using a very old version 2.x

What version specifically, and what version are you using now? Did this problem start occurring when upgrading to a new version?

**Reply by mattxerini** [reply]
If I recall correclty 2.6. Yes I am upgrading all the packages to the newest version and the errors started appearing with the newest changes.

**Reply by aruniverse** [reply]
2.x of the viewer pkgs or the the itwinjs-core pkgs?

**Reply by mattxerini** [reply]
2.6 of itwinjs-core

**Reply by aruniverse** [reply]
are you guys using the itwin viewer at all? Or did you roll your own viewer from an earlier sample?

Note, 2.6 is 3 yrs old, and there have been hundreds, if not thousands of changes since then. 
![image](https://user-images.githubusercontent.com/11051042/230116387-13050106-566c-4106-bfbd-9cf0fd8d0462.png)
The screenshot you've provided doesnt give us much to go off. Can you share a call stack? Share your code in a gh repo?

I would recommend taking a look at setting up a new viewer with the latest versions and see if you can migrate over.
https://github.com/iTwin/viewer/tree/master/packages/modules/cra-template-web-viewer#itwin-viewer-create-react-app-template-for-web

**Reply by mattxerini** [reply]
Yes, the iTwinJS was very old, hasn't been updated for years. 
We were using   "@bentley/itwin-viewer-react" which according to the doc has been removed and now I am using the @itwin/web-viewer-react"

**Reply by aruniverse** [reply]
Can you see if you can reproduce the access token issue on a new app using the above template i linked?

**Reply by mattxerini** [reply]
Yes, will try to create a new app now

**Reply by mattxerini** [reply]
Alright, I have been trying to follow your suggestion altough the CRA script isn't useful for me as what we intend to do it is slightly different. I will try to explain below bear with me.

So the way this needs to work, doesn't require the user to authenticate or interact with the viewer and itwin authentication steps, so we have implemented a backend logic that generate the access token using a SystemUser. Then this access token is passed to the frontend and then it will be passed to the API. 

We followed [AUTHORIZING SERVICE (MACHINE-TO-MACHINE)](https://developer.bentley.com/apis/overview/authorization/#authorizing-service-machine-to-machine)

<img width="848" alt="image" src="https://user-images.githubusercontent.com/123177586/230425189-13c72bed-b2a9-4418-845d-4ec0783439ed.png">

The frontend then has the following interface that pass the token to the api
<img width="941" alt="image" src="https://user-images.githubusercontent.com/123177586/230424536-2757045a-9167-4e43-b635-edb38ac9f55d.png">
<img width="936" alt="image" src="https://user-images.githubusercontent.com/123177586/230425564-f228d8f0-c23f-422c-9ab2-7f0d0f5d1a32.png">

As the version we used it was very old (as  I said 2.6), a few things have been changed. The classes FrontendAuthorizationClient, ClientRequestContext, AccessToken have been either removed or refactored. The first two has been removed and I cannot find what would be its substitute and the AccessToken has been refactored to a simple string.

My issue comes from the fact that whatever I do the token I passed (even a hardcoded one) to the api doesn't seems correct anymore and fails in various places.  Could fail as per initial message, due to the format or later when it tries to use it in the Authorization Header
<img width="742" alt="image" src="https://user-images.githubusercontent.com/123177586/230428825-1fe45b5c-43a3-46c1-9f37-ee702cd66858.png">


How could I implemented the interface we used to have to feed the token to the api? What sort of classes should we use and what sort of format the AccessToken string should look like?

**Reply by aruniverse** [reply]
Have you seen this [proxy viewer sample](https://github.com/calebmshafer/itwin-viewer-proxy)?

**Comment by mattxerini** [comment]
Good morning, I tried that implementation last week. 
![image](https://user-images.githubusercontent.com/123177586/231107980-3d47a5bb-4c63-4e96-aa07-e2330131a5ba.png)

Although as I mentioned already above the server side generated token response return a quite different response than what the api expect, or at least this is what the docs says. 
The getAccessTokenClientCredieantialGrant return an OAuth object, assuming we have to use those property to send the right format i.e typeToken + " " + accessToken. The second error I mentioned occur.

{"error":{"code":"HeaderNotFound","message":"Header Authorization was not found in the request. Access denied."}}

![image](https://user-images.githubusercontent.com/123177586/231133334-dc4e99f6-1f30-4507-822f-370268a7937a.png)
![image](https://user-images.githubusercontent.com/123177586/231133447-c6ccb254-c6ad-440d-add4-103e4bd5c1f6.png)

**Reply by aruniverse** [reply]
> The getAccessTokenClientCredieantialGrant return an OAuth object

Could you share what this object looks like?

**Reply by mattxerini** [reply]
{
  "accessToken": { generatedToken },
  "refreshToken": "null",
  "tokenType": "Bearer",
  "expiresIn": 3599,
  "scope": null
  "rawResponse": { stringify version of this entire object } 
  }

**Reply by aruniverse** [reply]
What does the content of the `generatedToken` look like?

**Reply by mattxerini** [reply]
something like this

ey....6rX7zJJ2kA

**Reply by aruniverse** [reply]
Ok good, thats the access token (string) the frontend expects. Your token url should return `Bearer ey....` as the access token.

**Reply by aruniverse** [reply]
You can use jwt.ms to decode your access token. Note the client id, which points to a spa auth workflow, and not an actual service (machine to machine) workflow. Client ids for services would begin with `service-` prefix.

**Reply by mattxerini** [reply]
The one I sent was from an example I created for testing. For security purpose I cannot share the other token, but I can assure you it is correct. I can also confirm the client ID has `service-` prefix you mentioned

**Reply by aruniverse** [reply]
If you're server is returning `Bearer ey....`, then im struggling to see why it'd fail with unexpected format.
Can you put a breakpoint on `accessToken.split` here? What is accessToken? Is it not the `Bearer ey...` string?
<img width="742" alt="image" src="https://user-images.githubusercontent.com/123177586/230428825-1fe45b5c-43a3-46c1-9f37-ee702cd66858.png">

**Reply by mattxerini** [reply]
The server return a OAtuh Object. Following the machine to machine workflow. 
<img width="1118" alt="image" src="https://user-images.githubusercontent.com/123177586/231214613-d0936abb-dc64-4a9c-8cec-c70825210ee0.png">

![image](https://user-images.githubusercontent.com/123177586/231214277-64c19eed-23c3-4924-b8d0-37cfdea086e6.png)

So you saying that we need to create the right token format manually using the returned object? 
That's fine, then after I pass the token to the api ` this.onAccessTokenChanged.raiseEvent(tokenString)` the error mention above occur. See the errors in the screenshot 401 and above unkown error in the AxiosRestClient.parse

**Reply by mattxerini** [reply, marked as answer]
Alright, thank you all for the replies. 
The following resolutions were confirmed by iTwin devs

The first issues (token format) was resolved using the OAuth object properties and create a proper Bearer Token format i.e `"Bearer ey...."`. So for future reference for anyone coming from a very old version and using Machine to Machine flow, you will have to create the format on the fly in your application using the OAuth object and pass it through the api, using something similiar to the `MyTokenServerAuthclient` (see screenshot above)

The Second issue which was a 401 in the `AxiosRestClient.parse` it was due to the incorrect uri used. As per documentation it seems you have to use the ims.bentley.com/connect/token

**Reply by aruniverse** [reply]
@mattxerini Matt, glad you were able to get things working!
